packages/storage/src/service.ts

```diff
@@ -16,55 +16,146 @@
+    }
+    const reference = new Reference(service, service._bucket!);
+    if (path != null) {
+      return refFromPath(reference, path);
+    } else {
+      return reference;
+    }
+  } else {
+    // ref is a Reference
+    if (typeof path === 'string') {
+      if (path.includes('..')) {
+        throw invalidArgument('`path` param cannot contain ".."');
+      }
+      return getChild(ref, path);
+    } else {
+      return ref;
+    }
+  }
+}
+
+export function ref(storage: StorageService, url?: string): Reference;
+export function ref(
+  storageOrRef: StorageService | Reference,
+  path?: string
+): Reference;
+/**
+ * Returns a storage Reference for the given url, or given path in the
+ * default bucket.
+ * @param serviceOrRef - `Storage` instance or storage `Reference`.
+ * @param pathOrUrlStorage - path, or URL. If empty, returns root reference (if Storage
+ * instance provided) or returns same reference (if Reference provided).
+ */
+export function ref(
+  serviceOrRef: StorageService | Reference,
+  pathOrUrl?: string
+): Reference | null {
+  if (pathOrUrl && isUrl(pathOrUrl)) {
+    if (serviceOrRef instanceof StorageService) {
+      return refFromURL(serviceOrRef, pathOrUrl);
+    } else {
+      throw invalidArgument(
+        'To use ref(service, url), the first argument must be a Storage instance.'
+      );
+    }
+  } else if (pathOrUrl !== null) {
+    if (
+      serviceOrRef instanceof StorageService ||
+      serviceOrRef instanceof Reference
+    ) {
+      return refFromPath(serviceOrRef, pathOrUrl);
+    } else {
+      throw invalidArgument(
+        'To use ref(serviceOrRef, path), the first argument must be a Storage' +
+          ' instance or Reference.'
+      );
+    }
+  } else {
+    // pathOrUrl param not provided
+    if (serviceOrRef instanceof StorageService) {
+      return refFromPath(serviceOrRef);
+    } else {
+      throw invalidArgument(
+        'To get the root reference, a Storage instance must be provided as' +
+          ' the param. Example: ref(storageInstance);'
+      );
+    }
+  }
+}
+
+/**
+ * A service that provides Firebase Storage Reference instances.
  * @param opt_url gs:// url to a custom Storage Bucket
- *
- * @struct
  */
 export class StorageService {
-  private app_: FirebaseApp | null;
-  private readonly bucket_: Location | null = null;
-  private readonly internals_: ServiceInternals;
-  private readonly authProvider_: Provider<FirebaseAuthInternalName>;
-  private readonly appId_: string | null = null;
-  private readonly pool_: XhrIoPool;
-  private readonly requests_: Set<Request<unknown>>;
-  private deleted_: boolean = false;
-  private maxOperationRetryTime_: number;
-  private maxUploadRetryTime_: number;
+  /**
+   * @internal
+   */
+  readonly _bucket: Location | null = null;
+  protected readonly _appId: string | null = null;
+  private readonly _requests: Set<Request<unknown>>;
+  /**
+   * @internal
+   */
+  private _deleted: boolean = false;
```

Comment:
private properties can only be accessed by the same class, so it doesn't need to be marked as internal.

Reply:
Oops.